### PR TITLE
[RFC: HTTP::Request] provide uri component accessors

### PR DESCRIPTION
**Pull request by waterlink**
Implements first option from this comment: https://github.com/manastech/crystal/pull/1505#issuecomment-141016552

For simplicity and smallest merge-able step, this does not implement any kind of smart wrapper over `#query`, instead it is just a string like all other components.

WDYT? 

/cc @jhass @asterite @waj

### CODE REVIEW

File: src/http/request.cr [hunk 1]
```diff
@@ -22,8 +22,45 @@ class HTTP::Request
     @cookies ||= Cookies.from_headers(headers)
   end
 
+  # Lazily parses and returns URI struct from request.
   def uri
-    URI.parse(@path)
+    (@uri ||= URI.parse(@resource)).not_nil!
+  end
```
Comment (jhass) [inline src/http/request.cr]: I wonder if this should take `@headers["Host"]` into account if present. It definitely should make sure the scheme is `"http"` I think, even though that may be misleading towards HTTPS requests.

**Comment by ysbaddaden** [comment]
What about delegating to uri, instead of having to sync both uri and local attributes?

**Comment by asterite** [comment]
What's a use case for having to modify a Request object once its constructed?

**Comment by jhass** [comment]
@asterite I could imagine passing it to something that may or may not change it prior sending it using HTTP::Client, but also some rewriting in a server middleware, say normalizing it from multiple routes to a canonical one.

File: src/http/request.cr (hunk 1 above)
Comment (waterlink) [inline src/http/request.cr]: Sounds good.

**Comment by waterlink** [comment]
@asterite, 

I didn't have any use case in mind. But now I think about something like proxy? (accept request as a server, mutate it and send it to another server).

I am actually happy to make it non-mutable (at least by obvious means), will simplify a lot of this PR, and for my purposes I would never even try to modify it.

You can always instantiate new request with adjusted attributes.

@ysbaddaden,

Agree, should delegate getters here. Everything can't be delegated, because of performance reasons - we don't want to parse `@resource` if `#uri` and derivatives never have been called.

**Comment by asterite** [comment]
Fair enough. The changes looks really nice, I like it that the URI object is only constructed if needed.

But there's this issue:

```crystal
require "http"

request = HTTP::Request.new "GET", "/"
pp request.resource #=> "/"

request.uri.path = "/hello"
pp request.resource #=> "/"
```

So maybe we can do what @ysbaddaden says, but also make `resource` be `@uri.try(&.to_s) || @resource`. That way we can also remove all the `path`, `path=`, `query`, `query=` and just do `request.uri.query`/`request.uri.query = ...`, leaving Request slimmer. What do you think?

In any case, I'd like to know what @waj thinks.

**Comment by asterite** [comment]
I mean this:

```crystal
  def resource
    @uri.try(&.to_s) || @resource
  end
```

and using `resource` instead of `@resource` in `to_s`.

**Comment by waterlink** [comment]
@asterite Nice catch with `@uri.try(&.to_s) || @resource` :)

**Comment by asterite** [comment]
Oh, and instead of `@uri.try(&.to_s)` it should probably be `@uri.try(&.full_path)`. I also wonder about `@headers["Host"]`, but I don't know what to do with that.

**Comment by waterlink** [comment]
So, I have added some fixes:

- Remove `fragment` related code - apparently `fragments` should not be in HTTP requests, they are client-only.
- Make `#path` and `#query` delegate to `#uri`.
- Remove `#path=(v)` and `#query=(v)`.
- Make `#resource` use `@uri` if it was already parsed, otherwise it is just initial `@resource`.
- Add specs that verify difference between `@uri.to_s` and `@uri.full_path`.

**Comment by waterlink** [comment]
+ Added logic for keeping header `Host` and `uri.{host,port}` in sync.

**Comment by waterlink** [comment]
Hmm, I am actually reluctant to revert the commit with `headers[Host] <-> uri.{host,port}` sync: because this behavior might be confusing and buggy:

```crystal
request.headers["Host"] = "john.example.org"
request.uri.host = "sarah.example.org"
request.to_io(io); io.to_s
# => "GET / HTTP/1.1\r\nHost: ?????.example.org\r\n\r\n"
```

**Comment by waterlink** [comment]
Reverted it. If you are interested how it was implemented, I leave a commit + revert there before final rebase for discussion.

**Comment by jhass** [comment]
If we update the path that way we should update the host too IMO

**Comment by waterlink** [comment]
@jhass Another way to do it is to completely hide `#uri` and provide delegate methods for path and query. In that case, there will not be other possibility to modify (or even read) `Host` other than through `#headers`.

**Comment by ysbaddaden** [comment]
Sounds good to me. As long as we have access to the different elements of `@resource`, we don't need to access `.uri`.

**Comment by waterlink** [comment]
Added an update again:

- Add `#path=` and `#query=` delegation to `#uri`
- Make `#uri` internal detail of `HTTP::Request` - as a `private` method

Why? - Exposing less data to the user of this class leaves less possibilities of "mis-usage" of it.

If you need to play around with `URI` object out of `HTTP::Request`, you can always build it from `#resource` yourself.

File: src/http/server/handlers/static_file_handler.cr [hunk 2]
```diff
@@ -5,9 +5,9 @@ class HTTP::StaticFileHandler < HTTP::Handler
   end
 
   def call(request)
-    file_path = @publicdir + request.path
+    file_path = @publicdir + request.resource
     if Dir.exists?(file_path)
-      HTTP::Response.new(200, directory_listing(request.path, file_path), HTTP::Headers{"Content-Type": "text/html"})
+      HTTP::Response.new(200, directory_listing(request.resource, file_path), HTTP::Headers{"Content-Type": "text/html"})
```
Comment (ysbaddaden) [inline src/http/server/handlers/static_file_handler.cr]: Minor detail: shouldn't it target the file at the request path, skipping the query params?

Comment (waterlink) [inline src/http/server/handlers/static_file_handler.cr]: You are right it should just be #path here

File: src/compiler/crystal/tools/browser.cr [hunk 3]
```diff
@@ -12,7 +12,7 @@ class Crystal::Browser
     while true
       server.accept do |sock|
         if request = HTTP::Request.from_io(sock)
-          html = browser.handle(request.path)
+          html = browser.handle(request.resource)
```
Comment (waterlink) [inline src/compiler/crystal/tools/browser.cr]: Same here, should have just been left as `#path`

File: src/http/server/handlers/static_file_handler.cr (hunk 2 above)
Comment (waterlink) [inline src/http/server/handlers/static_file_handler.cr]: Fixed

**Comment by jhass** [comment]
I think the current state is uncontroversial, I'd say time to squash it down and then we can further iterate in follow ups, if necessary :)

**Comment by ysbaddaden** [comment]
:+1:

**Comment by waterlink** [comment]
Rebased

**Comment by asterite** [comment]
I agree, this is the best solution :)

**Comment by jhass** [comment]
Thank you!

**Comment by Nephos** [comment]
Hello, I did not found the way to access to the scheme and port from the HTTP::Request.

Is something like the following patch seems acceptable ?

```patch
diff --git a/src/http/request.cr b/src/http/request.cr
index 5c005ee..65f321f 100644
--- a/src/http/request.cr
+++ b/src/http/request.cr
@@ -87,6 +87,15 @@ class HTTP::Request
     value
   end
 
+  delegate "scheme", uri
+  delegate "scheme=", uri
+
+  delegate "host", uri
+  delegate "host=", uri
+
+  delegate "port", uri
+  delegate "port=", uri
+
   private def uri
     (@uri ||= URI.parse(@resource)).not_nil!
   end
diff --git a/src/http/server/server.cr b/src/http/server/server.cr
index 427e85e..e055cc0 100644
--- a/src/http/server/server.cr
+++ b/src/http/server/server.cr
@@ -172,6 +172,16 @@ class HTTP::Server
         end
         break unless request
 
+        ifdef !without_openssl
+          ssl = @ssl
+        else
+          ssl = false
+        end
+
+        request.host = request.headers["Host"].split(":").first
+        request.port = (request.headers["Host"].split(":")[1]? || "80").to_i
+        request.scheme = ssl ? "https" : "http"
+
         response.version = request.version
         response.reset
         response.headers["Connection"] = "keep-alive" if request.keep_alive?
```

**Comment by jhass** [comment]
Why setters? They aren't used for anything.

**Comment by Nephos** [comment]
I thought to use it here
```crystal
# /src/http/server/server.cr#181
        request.host = request.headers["Host"].split(":").first
        request.port = (request.headers["Host"].split(":")[1]? || "80").to_i
        request.scheme = ssl ? "https" : "http"
```

because ``request = HTTP::Request.from_io(io)`` cannot set the scheme and port.

**Comment by Nephos** [comment]
Yes i agree, but there is now way to get the scheme and port in a HTTP::Request.
Maybe theses informations should be provided to the HTTP::Request constructor ?

Maybe should I open a new issue to discuss it ?

**Comment by jhass** [comment]
Perhaps, yes. It would also a good case to detail the motivation for being able to access those through it.

**Comment by asterite** [comment]
Can't you get that info from the HTTP client? The request itself in the protocol doesn't carry this information, I can't see why we need to add it.

**Pull request merged**